pydomcfg/domzgr/sco.py

```diff
@@ -0,0 +1,356 @@
+        # ds["hbatt"] = self._envlp
+        return self._merge_z3_and_e3(z3t, z3w, e3t, e3w)
+
+    def _check_stretch_par(self, psurf, pbott, alpha, efold, pbot2):
+        """
+        Check consistency of stretching parameters
+        """
+        if not (psurf and pbott):
+            if self._stretch == "sh94":
+                srf = "rn_theta"
+                bot = "rn_bb"
+            elif self._stretch == "md96":
+                srf = "rn_theta"
+                bot = "rn_thetb"
+            elif self._stretch == "sf12":
+                srf = "rn_zs"
+                bot = "rn_zb_a"
+            raise ValueError(
+                f"{srf} and {bot} MUST be set when using {self._stretch} stretching."
+            )
+
+        if self._stretch == "sf12":
+            if not (alpha and efold and pbot2):
+                raise ValueError(
+                    "rn_alpha, rn_efold and rn_zb_b MUST be set when using \
+                     sf12 stretching."
+                )
+
+    def _compute_env(self, depth: DataArray) -> DataArray:
+        """
+        Compute the envelope bathymetry surface by applying the
+        Martinho & Batteen (2006) smoothing algorithm to the
+        actual topography to reduce the maximum value of the slope
+        parameter
+                      r = abs(Hb-Ha) / (Ha+Hb)
+
+        where Ha and Hb are the depths of adjacent grid cells.
+        The maximum slope parameter is reduced to be <= rmax.
+
+        Reference:
+          *) Martinho & Batteen, Oce. Mod. 13(2):166-175, 2006.
+
+        Parameters
+        ----------
+        depth: DataArray
+            xarray DataArray of the 2D bottom topography
+            it MUST have only two dimensions
+        Returns
+        -------
+        DataArray
+            xarray DataArray of the 2D envelope bathymetry
+        """
+
+        if self._rmax:
+
+            lsm = self._lsm
+
+            # set first land point adjacent to a wet cell to
+            # min_dep as this needs to be included in smoothing
+            cst_lsm = lsm.rolling({dim: 3 for dim in lsm.dims}, min_periods=2).sum()
+            cst_lsm = cst_lsm.shift({dim: -1 for dim in lsm.dims})
+            cst_lsm = (cst_lsm > 0) & (lsm == 0)
+            zenv = depth.where(cst_lsm == 0, self._min_dep)
+
+            zenv = _smooth_MB06(zenv, self._rmax)
+            zenv = zenv.where(zenv > self._min_dep, self._min_dep)
+
+        return zenv
+
+    @property
+    def _sco_z3(self) -> Tuple[DataArray, ...]:
+        """Compute and return z3{t,w} for s-coordinates grids"""
+
+        grids = ("T", "W")
+        sigmas = self._sigmas
+        scosrf = self._envlp * 0.0  # unperturbed free-surface
+
+        both_z3 = []
+        for grid, sigma in zip(grids, sigmas):
+
+            if self._stretch:
+                # Stretched sco grid
+                su = -sigma
+                ss = self._stretch_sco(-sigma)
+                a1 = scosrf
+                a2 = 0.0
```

Comment:
Although I'm not sure why, `mypy` doesn't like `a2` being `float` in one case and `DataArray` in the other.
Let's have a look at it when we tidy up this PR.
```suggestion
                a2 = DataArray((0.0))  # TODO: Why can't use float here?
```